File: municipal_rosters.py
"""Helpers for downloading Summit County municipal STR license rosters."""

from __future__ import annotations

import json
import logging
import os
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Dict, Iterable, List, Optional, Sequence

from arcgis.features import FeatureLayer
from arcgis.gis import GIS


LOGGER = logging.getLogger(__name__)


def _normalise_schedule_number(value: Any) -> Optional[str]:
    if value is None:
        return None
    text = str(value).strip()
    if not text:
        return None
    return text.upper()


_STATUS_ALIASES = {
    "APPROVED": "active",
    "ACTIVE": "active",
    "ISSUED": "active",
    "CURRENT": "active",
    "GOOD STANDING": "active",
    "IN GOOD STANDING": "active",
    "RENEWED": "active",
    "PAID": "active",
    "PENDING": "pending",
    "UNDER REVIEW": "pending",
    "IN PROCESS": "pending",
    "EXPIRED": "expired",
    "INACTIVE": "inactive",
    "SUSPENDED": "inactive",
    "REVOKED": "revoked",
    "DENIED": "revoked",
    "CANCELLED": "revoked",
    "CANCELED": "revoked",
}


def _normalise_status(value: Any) -> str:
    if value is None:
        return "unknown"
    text = str(value).strip()
    if not text:
        return "unknown"
    upper = text.upper()
    for key, alias in _STATUS_ALIASES.items():
        if key in upper:
            return alias
    return "unknown"


def _parse_date(value: Any) -> Optional[str]:
    if value in (None, ""):
        return None
    if isinstance(value, datetime):
        return value.date().isoformat()
    if isinstance(value, (int, float)):
        if value > 1e12:
            try:
                return datetime.utcfromtimestamp(value / 1000).date().isoformat()
            except (OverflowError, ValueError):
                return None
        if value > 1e9:
            try:
                return datetime.utcfromtimestamp(value).date().isoformat()
            except (OverflowError, ValueError):
                return None
    try:
        parsed = datetime.fromisoformat(str(value))
        return parsed.date().isoformat()
    except ValueError:
        pass
    try:
        parsed = datetime.strptime(str(value), "%m/%d/%Y")
        return parsed.date().isoformat()
    except ValueError:
        pass
    try:
        parsed = datetime.strptime(str(value), "%Y-%m-%d %H:%M:%S")
        return parsed.date().isoformat()
    except ValueError:
        return None


@dataclass
class MunicipalRosterSource:
    """Configuration describing how to query a municipal STR roster."""

    municipality: str
    layer_url: str
    schedule_field: str
    license_id_field: str
    status_field: str
    expiration_field: Optional[str] = None
    updated_field: Optional[str] = None
    where: str = "1=1"
    out_fields: Sequence[str] = field(default_factory=lambda: ["*"])
    detail_url_template: Optional[str] = None

    def build_query(self) -> Dict[str, Any]:
        out_fields = self.out_fields if self.out_fields else ["*"]
        return {
            "where": self.where or "1=1",
            "outFields": ",".join(out_fields),
            "returnGeometry": json.dumps(False),
        }


@dataclass
class MunicipalLicenseRecord:
    municipality: str
    schedule_number: str
    license_id: str
    status: str
    normalized_status: str
    expiration_date: Optional[str]
    updated_at: Optional[str]
    detail_url: Optional[str]
    raw: Dict[str, Any]

    def to_dict(self) -> Dict[str, Any]:
        return {
            "municipality": self.municipality,
            "schedule_number": self.schedule_number,
            "municipal_license_id": self.license_id,
            "status": self.status,
            "normalized_status": self.normalized_status,
            "expiration_date": self.expiration_date,
            "updated_at": self.updated_at,
            "detail_url": self.detail_url,
            "raw": self.raw,
        }


def _iter_features(layer: FeatureLayer, source: MunicipalRosterSource) -> Iterable[Dict[str, Any]]:
    page_size = getattr(getattr(layer, "properties", None), "maxRecordCount", 1000) or 1000
    offset = 0
    while True:
        payload = layer.query(
            where=source.where or "1=1",
            out_fields=",".join(source.out_fields or ["*"]),
            return_geometry=False,
            result_offset=offset,
            result_record_count=page_size,
        ).to_dict()
        features = payload.get("features", [])
        if not features:
            break
        for feature in features:
            attributes = feature.get("attributes") or {}
            yield attributes
        if len(features) < page_size:
            break
        offset += page_size


def _build_detail_url(source: MunicipalRosterSource, attributes: Dict[str, Any]) -> Optional[str]:
    template = source.detail_url_template
    if not template:
        return None
    try:
        return template.format(**attributes)
    except KeyError:
        return None


def _extract_record(source: MunicipalRosterSource, attributes: Dict[str, Any]) -> Optional[MunicipalLicenseRecord]:
    schedule_value = attributes.get(source.schedule_field)
    schedule_number = _normalise_schedule_number(schedule_value)
    if not schedule_number:
        return None

    license_value = attributes.get(source.license_id_field)
    license_id = str(license_value).strip() if license_value not in (None, "") else None
    if not license_id:
        return None

    status_value = attributes.get(source.status_field)
    status = str(status_value).strip() if status_value not in (None, "") else "Unknown"
    normalized_status = _normalise_status(status_value)

    expiration_value = attributes.get(source.expiration_field) if source.expiration_field else None
    expiration_date = _parse_date(expiration_value)

    updated_value = attributes.get(source.updated_field) if source.updated_field else None
    updated_at = _parse_date(updated_value)

    detail_url = _build_detail_url(source, attributes)

    return MunicipalLicenseRecord(
        municipality=source.municipality,
        schedule_number=schedule_number,
        license_id=license_id,
        status=status or "Unknown",
        normalized_status=normalized_status,
        expiration_date=expiration_date,
        updated_at=updated_at,
        detail_url=detail_url,
        raw=dict(attributes),
    )


def _load_source_overrides() -> Dict[str, MunicipalRosterSource]:
    overrides_path = os.getenv("SUMMIT_MUNICIPAL_ROSTERS")
    if not overrides_path:
        return {}
    try:
        with open(overrides_path, "r", encoding="utf-8") as fh:
            payload = json.load(fh)
    except OSError as exc:  # pragma: no cover - defensive
        LOGGER.warning("Unable to read municipal roster overrides: %s", exc)
        return {}
    except json.JSONDecodeError as exc:  # pragma: no cover - defensive
        LOGGER.warning("Invalid JSON in municipal roster overrides: %s", exc)
        return {}

    overrides: Dict[str, MunicipalRosterSource] = {}
    for entry in payload:
        try:
            source = MunicipalRosterSource(
                municipality=entry["municipality"],
                layer_url=entry["layer_url"],
                schedule_field=entry["schedule_field"],
                license_id_field=entry["license_id_field"],
                status_field=entry["status_field"],
                expiration_field=entry.get("expiration_field"),
                updated_field=entry.get("updated_field"),
                where=entry.get("where", "1=1"),
                out_fields=entry.get("out_fields", ["*"]),
                detail_url_template=entry.get("detail_url_template"),
            )
        except KeyError as exc:
            LOGGER.warning("Skipping municipal roster override missing key %s", exc)
            continue
        overrides[source.municipality.lower()] = source
    return overrides


DEFAULT_MUNICIPAL_SOURCES: Dict[str, MunicipalRosterSource] = {
    "breckenridge": MunicipalRosterSource(
        municipality="Breckenridge",
        layer_url=os.getenv(
            "BRECKENRIDGE_ROSTER_URL",
            "https://services1.arcgis.com/DbqCQ5IIGIgjLU4g/arcgis/rest/services/STR_Licenses_Public/FeatureServer/0",
        ),
        schedule_field=os.getenv("BRECKENRIDGE_SCHEDULE_FIELD", "SCHEDULE_NUM"),
        license_id_field=os.getenv("BRECKENRIDGE_LICENSE_FIELD", "LICENSE_NO"),
        status_field=os.getenv("BRECKENRIDGE_STATUS_FIELD", "STATUS"),
        expiration_field=os.getenv("BRECKENRIDGE_EXPIRATION_FIELD", "EXPIRATION"),
        updated_field=os.getenv("BRECKENRIDGE_UPDATED_FIELD", "LAST_UPDATE"),
        detail_url_template=os.getenv(
            "BRECKENRIDGE_LICENSE_URL_TEMPLATE",
            "https://www.townofbreckenridge.com/str/{LICENSE_NO}",
        ),
    ),
    "frisco": MunicipalRosterSource(
        municipality="Frisco",
        layer_url=os.getenv(
            "FRISCO_ROSTER_URL",
            "https://services7.arcgis.com/r0nAYG7DmzNoKGbT/arcgis/rest/services/Frisco_STR_Licenses/FeatureServer/0",
        ),
        schedule_field=os.getenv("FRISCO_SCHEDULE_FIELD", "SCHEDULE"),
        license_id_field=os.getenv("FRISCO_LICENSE_FIELD", "LICENSE_NO"),
        status_field=os.getenv("FRISCO_STATUS_FIELD", "STATUS"),
        expiration_field=os.getenv("FRISCO_EXPIRATION_FIELD", "EXPIRATION"),
        updated_field=os.getenv("FRISCO_UPDATED_FIELD", "LASTUPDATED"),
    ),
    "dillon": MunicipalRosterSource(
        municipality="Dillon",
        layer_url=os.getenv(
            "DILLON_ROSTER_URL",
            "https://services7.arcgis.com/4W0wSZ3KFcuX39pB/arcgis/rest/services/Dillon_STR_Licenses/FeatureServer/0",
        ),
        schedule_field=os.getenv("DILLON_SCHEDULE_FIELD", "SCHEDULE"),
        license_id_field=os.getenv("DILLON_LICENSE_FIELD", "LICENSE_NO"),
        status_field=os.getenv("DILLON_STATUS_FIELD", "STATUS"),
        expiration_field=os.getenv("DILLON_EXPIRATION_FIELD", "EXPIRATION"),
        updated_field=os.getenv("DILLON_UPDATED_FIELD", "LAST_UPDATED"),
    ),
    "silverthorne": MunicipalRosterSource(
        municipality="Silverthorne",
        layer_url=os.getenv(
            "SILVERTHORNE_ROSTER_URL",
            "https://services7.arcgis.com/p0mEetxHUAZJr0qG/arcgis/rest/services/Silverthorne_STR_Licenses/FeatureServer/0",
        ),
        schedule_field=os.getenv("SILVERTHORNE_SCHEDULE_FIELD", "SCHEDULE"),
        license_id_field=os.getenv("SILVERTHORNE_LICENSE_FIELD", "LICENSE_NO"),
        status_field=os.getenv("SILVERTHORNE_STATUS_FIELD", "STATUS"),
        expiration_field=os.getenv("SILVERTHORNE_EXPIRATION_FIELD", "EXPIRATION"),
        updated_field=os.getenv("SILVERTHORNE_UPDATED_FIELD", "LAST_MODIFIED"),
    ),
}


def load_municipal_sources() -> Dict[str, MunicipalRosterSource]:
    sources = dict(DEFAULT_MUNICIPAL_SOURCES)
    overrides = _load_source_overrides()
    sources.update(overrides)
    return sources


def fetch_municipal_rosters(
    gis: GIS,
    *,
    sources: Optional[Dict[str, MunicipalRosterSource]] = None,
    logger: Optional[logging.Logger] = None,
) -> List[MunicipalLicenseRecord]:
    """Fetch and normalise STR licenses for all configured municipalities."""

    if sources is None:
        sources = load_municipal_sources()

    records: List[MunicipalLicenseRecord] = []

    for key, source in sources.items():
        if not source.layer_url:
            if logger:
                logger.warning("Municipal roster '%s' is missing a layer URL; skipping.", key)
            continue
        layer = FeatureLayer(source.layer_url, gis=gis)
        fetched = 0
        for attributes in _iter_features(layer, source):
            record = _extract_record(source, attributes)
            if record is None:
                continue
            records.append(record)
            fetched += 1
        if logger:
            logger.info("Fetched %s municipal STR licenses for %s.", fetched, source.municipality)

    return records


def dump_municipal_rosters(
    gis: GIS,
    *,
    output_path: Optional[str] = None,
    logger: Optional[logging.Logger] = None,
) -> Dict[str, Any]:
    records = fetch_municipal_rosters(gis, logger=logger)
    payload = {
        "generated_at": datetime.utcnow().replace(microsecond=0).isoformat() + "Z",
        "record_count": len(records),
        "records": [record.to_dict() for record in records],
    }
    if output_path:
        with open(output_path, "w", encoding="utf-8") as fh:
            json.dump(payload, fh, indent=2)
    return payload


__all__ = [
    "MunicipalRosterSource",
    "MunicipalLicenseRecord",
    "fetch_municipal_rosters",
    "dump_municipal_rosters",
    "load_municipal_sources",
]

File: webapp/arcgis-webapp/scripts/municipalRosters.mjs
import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';

const STATUS_ALIASES = new Map([
  ['APPROVED', 'active'],
  ['ACTIVE', 'active'],
  ['ISSUED', 'active'],
  ['CURRENT', 'active'],
  ['GOOD STANDING', 'active'],
  ['IN GOOD STANDING', 'active'],
  ['RENEWED', 'active'],
  ['PAID', 'active'],
  ['PENDING', 'pending'],
  ['UNDER REVIEW', 'pending'],
  ['IN PROCESS', 'pending'],
  ['EXPIRED', 'expired'],
  ['INACTIVE', 'inactive'],
  ['SUSPENDED', 'inactive'],
  ['REVOKED', 'revoked'],
  ['DENIED', 'revoked'],
  ['CANCELLED', 'revoked'],
  ['CANCELED', 'revoked'],
]);

const DEFAULT_REFERER =
  process.env.SUMMIT_ARCGIS_REFERER ??
  process.env.SUMMIT_MUNICIPAL_REFERER ??
  'https://experience.arcgis.com/experience/706a6886322445479abadb904db00bc0/';

const STATUS_USER_AGENT = 'arcgis-webapp-metrics/1.0';

function buildArcgisHeaders() {
  return {
    Referer: DEFAULT_REFERER,
    'User-Agent': STATUS_USER_AGENT,
  };
}

const DEFAULT_SOURCES = [
  {
    key: 'breckenridge',
    municipality: 'Breckenridge',
    layerUrl:
      process.env.BRECKENRIDGE_ROSTER_URL ??
      'https://services6.arcgis.com/dmNYNuTJZDtkcRJq/arcgis/rest/services/STR_Licenses_October_2025_public_view_layer/FeatureServer/0',
    scheduleField: process.env.BRECKENRIDGE_SCHEDULE_FIELD ?? 'HC_RegistrationsOriginalCleaned',
    licenseIdField: process.env.BRECKENRIDGE_LICENSE_FIELD ?? 'HC_RegistrationsOriginalClean_1',
    statusField: process.env.BRECKENRIDGE_STATUS_FIELD ?? 'HC_RegistrationsOriginalClea_43',
    expirationField: process.env.BRECKENRIDGE_EXPIRATION_FIELD ?? null,
    updatedField: process.env.BRECKENRIDGE_UPDATED_FIELD ?? null,
    where: '1=1',
    outFields: ['*'],
    detailUrlTemplate:
      process.env.BRECKENRIDGE_LICENSE_URL_TEMPLATE ??
      'https://www.townofbreckenridge.com/str/{HC_RegistrationsOriginalClean_1}',
  },
  {
    key: 'unincorporated_summit_county',
    municipality: 'Unincorporated Summit County',
    layerUrl:
      process.env.UNINCORPORATED_SUMMIT_COUNTY_ROSTER_URL ??
      'https://services6.arcgis.com/dmNYNuTJZDtkcRJq/arcgis/rest/services/STR_Licenses_October_2025_public_view_layer/FeatureServer/0',
    scheduleField:
      process.env.UNINCORPORATED_SUMMIT_COUNTY_SCHEDULE_FIELD ?? 'HC_RegistrationsOriginalCleaned',
    licenseIdField:
      process.env.UNINCORPORATED_SUMMIT_COUNTY_LICENSE_FIELD ?? 'HC_RegistrationsOriginalClean_1',
    statusField:
      process.env.UNINCORPORATED_SUMMIT_COUNTY_STATUS_FIELD ?? 'HC_RegistrationsOriginalClea_43',
    expirationField: process.env.UNINCORPORATED_SUMMIT_COUNTY_EXPIRATION_FIELD ?? null,
    updatedField: process.env.UNINCORPORATED_SUMMIT_COUNTY_UPDATED_FIELD ?? null,
    where: '1=1',
    outFields: ['*'],
  },
];

function readJsonFile(filePath) {
  try {
    const resolved = path.resolve(filePath);
    if (!fs.existsSync(resolved)) {
      return null;
    }
    const raw = fs.readFileSync(resolved, 'utf8');
    return JSON.parse(raw);
  } catch (error) {
    return null;
  }
}

function loadSourceOverrides() {
  const overridesPath = process.env.SUMMIT_MUNICIPAL_ROSTERS;
  if (!overridesPath) {
    return [];
  }
  const payload = readJsonFile(overridesPath);
  if (!Array.isArray(payload)) {
    return [];
  }

  return payload
    .map((entry) => ({
      key: String(entry.municipality ?? '').toLowerCase(),
      municipality: entry.municipality,
      layerUrl: entry.layer_url,
      scheduleField: entry.schedule_field,
      licenseIdField: entry.license_id_field,
      statusField: entry.status_field,
      expirationField: entry.expiration_field ?? null,
      updatedField: entry.updated_field ?? null,
      where: entry.where ?? '1=1',
      outFields: Array.isArray(entry.out_fields) && entry.out_fields.length > 0 ? entry.out_fields : ['*'],
      detailUrlTemplate: entry.detail_url_template ?? null,
    }))
    .filter((entry) =>
      Boolean(entry.key && entry.municipality && entry.layerUrl && entry.scheduleField && entry.licenseIdField && entry.statusField),
    );
}

function loadMunicipalSources() {
  const overrides = loadSourceOverrides();
  const sources = new Map();
  for (const source of DEFAULT_SOURCES) {
    if (source.layerUrl) {
      sources.set(source.key, { ...source });
    }
  }
  for (const override of overrides) {
    sources.set(override.key, override);
  }
  return Array.from(sources.values());
}

function normaliseScheduleNumber(value) {
  if (value === null || value === undefined) {
    return null;
  }
  const text = String(value).trim();
  if (!text) {
    return null;
  }
  return text.toUpperCase();
}

function normaliseStatus(value) {
  if (value === null || value === undefined) {
    return 'unknown';
  }
  const text = String(value).trim();
  if (!text) {
    return 'unknown';
  }
  const upper = text.toUpperCase();
  for (const [key, alias] of STATUS_ALIASES.entries()) {
    if (upper.includes(key)) {
      return alias;
    }
  }
  return 'unknown';
}

function parseDate(value) {
  if (value === null || value === undefined || value === '') {
    return null;
  }

  if (value instanceof Date) {
    if (Number.isNaN(value.getTime())) {
      return null;
    }
    return value.toISOString().slice(0, 10);
  }

  if (typeof value === 'number' && Number.isFinite(value)) {
    if (value > 1e12) {
      const date = new Date(value);
      return Number.isNaN(date.getTime()) ? null : date.toISOString().slice(0, 10);
    }
    if (value > 1e9) {
      const date = new Date(value * 1000);
      return Number.isNaN(date.getTime()) ? null : date.toISOString().slice(0, 10);
    }
  }

  if (typeof value === 'string') {
    const trimmed = value.trim();
    if (!trimmed) {
      return null;
    }

    const isoCandidate = new Date(trimmed);
    if (!Number.isNaN(isoCandidate.getTime())) {
      return isoCandidate.toISOString().slice(0, 10);
    }

    const mmddyyyy = trimmed.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})$/);
    if (mmddyyyy) {
      const [_, month, day, year] = mmddyyyy;
      const date = new Date(Date.UTC(Number(year), Number(month) - 1, Number(day)));
      if (!Number.isNaN(date.getTime())) {
        return date.toISOString().slice(0, 10);
      }
    }

    const ymdHms = trimmed.match(/^(\d{4}-\d{2}-\d{2})\s+(\d{2}:\d{2}:\d{2})$/);
    if (ymdHms) {
      const [_, datePart, timePart] = ymdHms;
      const date = new Date(`${datePart}T${timePart}Z`);
      if (!Number.isNaN(date.getTime())) {
        return date.toISOString().slice(0, 10);
      }
    }
  }

  return null;
}

function buildDetailUrl(template, attributes) {
  if (!template) {
    return null;
  }
  try {
    return template.replace(/\{([^{}]+)\}/g, (_, key) => {
      const value = attributes?.[key];
      return value === null || value === undefined ? '' : String(value);
    });
  } catch (error) {
    return null;
  }
}

function buildRecordId(municipality, licenseId) {
  const key = `${municipality}::${licenseId}`;
  return crypto.createHash('sha256').update(key).digest('hex');
}

async function fetchFeaturePage(source, offset, limit) {
  const params = new URLSearchParams({
    f: 'json',
    where: source.where ?? '1=1',
    outFields: Array.isArray(source.outFields) && source.outFields.length > 0 ? source.outFields.join(',') : '*',
    returnGeometry: 'false',
    resultOffset: String(offset),
    resultRecordCount: String(limit),
  });

  const url = `${source.layerUrl.replace(/\/?$/, '')}/query?${params.toString()}`;
  const response = await fetch(url, { headers: buildArcgisHeaders() });
  if (!response.ok) {
    throw new Error(`Failed to fetch municipal roster ${source.municipality}: ${response.status} ${response.statusText}`);
  }

  const payload = await response.json();
  if (payload?.error) {
    const message = payload.error.message ?? 'Unknown ArcGIS error';
    const details = Array.isArray(payload.error.details) ? payload.error.details.join('; ') : '';
    throw new Error(`ArcGIS error for ${source.municipality}: ${message}${details ? ` (${details})` : ''}`);
  }

  const features = Array.isArray(payload?.features) ? payload.features : [];
  return { features, exceeded: Boolean(payload?.exceededTransferLimit) && features.length >= limit };
}

function toPlainObject(value) {
  if (!value || typeof value !== 'object') {
    return {};
  }
  try {
    return JSON.parse(JSON.stringify(value));
  } catch (error) {
    return {};
  }
}

function extractRecord(source, attributes) {
  const scheduleNumber = normaliseScheduleNumber(attributes?.[source.scheduleField]);
  if (!scheduleNumber) {
    return null;
  }

  const rawLicenseId = attributes?.[source.licenseIdField];
  const licenseId = rawLicenseId === null || rawLicenseId === undefined ? null : String(rawLicenseId).trim();
  if (!licenseId) {
    return null;
  }

  const rawStatus = attributes?.[source.statusField];
  const status = rawStatus === null || rawStatus === undefined ? 'Unknown' : String(rawStatus).trim() || 'Unknown';
  const normalizedStatus = normaliseStatus(rawStatus);

  const expirationValue = source.expirationField ? attributes?.[source.expirationField] : null;
  const expirationDate = parseDate(expirationValue);

  const updatedValue = source.updatedField ? attributes?.[source.updatedField] : null;
  const updatedDate = parseDate(updatedValue);

  const detailUrl = buildDetailUrl(source.detailUrlTemplate, attributes);

  return {
    id: buildRecordId(source.municipality, licenseId),
    municipality: source.municipality,
    scheduleNumber,
    licenseId,
    status,
    normalizedStatus,
    expirationDate,
    updatedDate,
    detailUrl,
    raw: toPlainObject(attributes ?? {}),
  };
}

async function fetchSourceRecords(source, logger) {
  const pageSize = 2000;
  const results = [];
  let offset = 0;

  while (true) {
    const { features, exceeded } = await fetchFeaturePage(source, offset, pageSize);
    for (const feature of features) {
      const attributes = feature?.attributes ?? feature ?? {};
      const record = extractRecord(source, attributes);
      if (record) {
        results.push(record);
      }
    }

    if (!exceeded || features.length === 0) {
      break;
    }
    offset += pageSize;
  }

  logger?.info?.(
    `[metrics] Fetched ${results.length.toLocaleString()} municipal STR licenses for ${source.municipality}.`,
  );

  return results;
}

export async function fetchMunicipalLicenseRecords(logger = console) {
  const sources = loadMunicipalSources();
  const records = [];
  const failures = [];

  for (const source of sources) {
    if (!source.layerUrl) {
      logger?.warn?.(
        `[metrics] Municipal roster "${source.key}" is missing a layer URL; skipping.`,
      );
      continue;
    }

    try {
      const sourceRecords = await fetchSourceRecords(source, logger);
      records.push(...sourceRecords);
    } catch (error) {
      failures.push({ source, error });
      logger?.error?.(`Failed to fetch municipal roster for ${source.municipality}: ${error?.message ?? error}`);
    }
  }

  if (records.length === 0 && failures.length === sources.length && sources.length > 0) {
    const details = failures
      .map(({ source, error }) => `${source.municipality}: ${error?.message ?? error}`)
      .join('; ');
    throw new Error(`Failed to fetch municipal rosters from ArcGIS (${details})`);
  }

  return records;
}

export function summariseMunicipalLicenses(records) {
  const municipalities = new Set(records.map((record) => record.municipality));
  return {
    total: records.length,
    municipalities: municipalities.size,
  };
}

export function normaliseMunicipalRecordForSupabase(record) {
  const sourceUpdatedAtIso = record.updatedDate
    ? `${record.updatedDate}T00:00:00Z`
    : null;

  return {
    id: record.id,
    schedule_number: record.scheduleNumber,
    municipality: record.municipality,
    municipal_license_id: record.licenseId,
    status: record.status,
    normalized_status: record.normalizedStatus,
    expiration_date: record.expirationDate,
    source_updated_at: sourceUpdatedAtIso,
    detail_url: record.detailUrl ?? null,
    raw: record.raw,
  };
}

export function groupLicensesBySchedule(records) {
  const bySchedule = new Map();
  for (const record of records) {
    const bucket = bySchedule.get(record.scheduleNumber) ?? [];
    bucket.push(record);
    bySchedule.set(record.scheduleNumber, bucket);
  }
  return bySchedule;
}
